### PR TITLE
Implement sending commands to solver in incremental SMT2 backend

### PR DESCRIPTION
**Pull request by thomasspriggs**
This PR includes implementation of sending commands to solver in incremental SMT2 backend, based on the exprts recieved. This is sufficient to get a sat/unsat response from the solver in sufficiently simple cases. However parsing the result from the solver is not included in this PR and will be in a follow up PR.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] N/A - Non claimed. ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp [hunk 1]
```diff
@@ -12,6 +13,9 @@ smt2_incremental_decision_proceduret::smt2_incremental_decision_proceduret(
     number_of_solver_calls{0},
     solver_process{split_string(solver_command, ' ', false, true)}
 {
+  send_to_solver(smt_set_option_commandt{smt_option_produce_modelst{true}});
+  send_to_solver(smt_set_logic_commandt{
+    smt_logic_quantifier_free_uninterpreted_functions_bit_vectorst{}});
```
Comment (TGWDB) [inline src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp]: Do we want to send these at solver construction time, or do we want to send/update these per query? I don't have a good answer now, but there was comment on another issue/PR that the logic could vary for different calls/inputs. Also do we want to be able to change these later (and is this allowed/easy to do)?

**Review by TGWDB — APPROVED** [review]

**Review by NlightNFotis — APPROVED** [review]
LGTM

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp [hunk 2]
```diff
@@ -8,10 +8,12 @@
...
 
 smt2_incremental_decision_proceduret::smt2_incremental_decision_proceduret(
-  std::string _solver_command)
-  : solver_command{std::move(_solver_command)},
+  std::string _solver_command,
+  message_handlert &message_handler)
+  : solver_command(std::move(_solver_command)),
     number_of_solver_calls{0},
-    solver_process{split_string(solver_command, ' ', false, true)}
+    solver_process{split_string(solver_command, ' ', false, true)},
```
Comment (peterschrammel) [inline src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp]: The split_string looks brittle.

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp (hunk 1 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp]: According to section 4.1 of the SMTLIB standard version 2.6, the `set-logic` command can only be issued whilst the solver is in "Start mode" and the only way to return to this mode is to issue the `reset` command. The actions of the `reset` command include undefining all functions and assertions we have sent to the solver. Therefore if we were reseting and setting different logics for different rounds of solving, then we would no longer be carrying out incremental solving.

TLDR; We should be sending these commands at construction time.

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp [hunk 3]
```diff
@@ -2,11 +2,93 @@
...
+    {
+      dependent_expressions.insert(expr_node);
+    }
+  });
+  return dependent_expressions;
+}
+
+/// \brief Defines any functions which \p expr depends on, which have not yet
+///   been defined, along with their dependencies in turn.
+void smt2_incremental_decision_proceduret::define_dependent_functions(
```
Comment (peterschrammel) [inline src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:43]: Would be great to unit-test this somehow.

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp (hunk 2 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp]: The arguments are provided to `cbmc` in space separated form. Therefore they currently need to be split in order separate them. Any suggestions for making this less brittle?

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp (hunk 3 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:43]: I have added unit tests of `smt2_incremental_decision_proceduret` these tests indirectly test the functionality of this `define_dependent_functions` function, via the `handle` and `set_to` functions of the public interface.

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp [hunk 4]
```diff
@@ -67,21 +67,21 @@ void smt2_incremental_decision_proceduret::define_dependent_functions(
     {
       const irep_idt &identifier = symbol_expr->get_identifier();
       const symbolt *symbol = nullptr;
-      if(!ns.lookup(identifier, symbol) && !symbol->value.is_nil())
+      if(ns.lookup(identifier, symbol) || symbol->value.is_nil())
```
Comment (NlightNFotis) [inline src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:73]: Are we sure this change is valid? It doesn't look like that to me.

[De Morgan's law in this particular instance would be a bi-implication](https://en.wikipedia.org/wiki/De_Morgan%27s_laws):

```
¬(A ∨ B) ⇔ ¬A ∧ ¬B
```

Assuming in your case `A : ns.lookup()` and `B : symbol->value.is_nil()` then I believe this should have been translated to `(!(ns.lookup() || symbol->value.is_nil()))`.

Comment (NlightNFotis) [reply src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:73]: Yeah, I just verified this is with a quick run of the truth table:

`A : ns.lookup()`
`B : symbol->value.is_nil()`
`Original : !ns.lookup() && !symbol->value.is_nil()`
`Refactored : ns.lookup() || symbol->value.is_nil()`

| A  | B | Original | Refactored |
|---|---|--------|-------------|
| F  | F  |      T      |         F           |
| F  | T  |       F     |         T           |
| T  | F  |       F     |         T           |
| T  | T  |       F     |        T            |

**Review by NlightNFotis — CHANGES_REQUESTED** [review]
The rest of the stuff looks good, but the refactoring of the condition expression changed its semantics.

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp (hunk 4 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:73]: Yes. The outcome of the logic you present is the opposite way round. That final not was handled by flipping the contents of the code blocks rather than adding the `not` operation. Therefore the outcome is still the same, which is why the tests still pass.

**Review by NlightNFotis — APPROVED** [review]

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp [hunk 5]
```diff
@@ -2,16 +2,133 @@
...
+void smt2_incremental_decision_proceduret::define_dependent_functions(
+  const exprt &expr)
+{
+  std::unordered_set<exprt, irep_hash> seen_expressions =
+    make_range(expression_identifiers)
+      .map([](const std::pair<exprt, smt_identifier_termt> &expr_identifier) {
+        return expr_identifier.first;
+      });
+  std::stack<exprt> to_be_defined;
+  const auto dependencies_needed = [&](const exprt &expr) {
```
Comment (chrisr-diffblue) [inline src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:52]: Any particular reason why this is a lambda, rather than a helper function, such as a hypothetical `push_dependencies_needed(exprt, std::stack<exprt>)` which might be more descriptive? Either way, I'd be tempted to change the name to make its behaviour more explicit... At the moment, it superficially looks like a call to `dependencies_needed(e)` would be state-free - but actually its changing the state of the method via the stack of exprts... I'd make that behaviour explicit in the name for ease of code reading...

File: src/solvers/smt2_incremental/smt_terms.cpp [hunk 6]
```diff
@@ -50,7 +50,7 @@ bool smt_bool_literal_termt::value() const
 
 static bool is_valid_smt_identifier(irep_idt identifier)
 {
-  static const std::regex valid{R"(^[^\|\\]*$)"};
+  static const std::regex valid{R"(^[^\|]*$)"};
```
Comment (chrisr-diffblue) [inline src/solvers/smt2_incremental/smt_terms.cpp:53]: Minor quibble - but for the more hieroglyphic regex's like this its nice to include a comment showing an/some examples of what the regex is attempting to match.

**Review by chrisr-diffblue — APPROVED** [review]
Nothing blocking - though I've left a couple of small comments that might be worth considering.

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/smt2_incremental/smt_terms.cpp (hunk 6 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_terms.cpp:53]: I agree this would be better with a comment. I didn't add one when I first wrote this, as it is relatively short and I am relatively comfortable with regex. I'll try and remember to comment this in a follow-up PR.

File: src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp (hunk 5 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:52]: Because it was the quickest way for me to de-duplicate this code in the 2 places it is needed. A separate function would also have needed an additional parameter. I agree that the name I assigned to this lambda isn't necessarily the best. I think doing some pair programming together could be helpful in trying to come up with a more readable implementation.
